hazelcast/src/hazelcast/client/compact.cpp

```diff
@@ -0,0 +1,678 @@
+            BOOST_THROW_EXCEPTION(exception::hazelcast_serialization(
+              "default_compact_writer",
+              (boost::format("Invalid field type %1% for %2%") % field_name %
+               schema_)
+                .str()));
+        }
+    }
+    return iterator->second;
+}
+
+void
+default_compact_writer::write_offsets(size_t data_length)
+{
+    if (data_length < offset_reader::BYTE_OFFSET_READER_RANGE) {
+        for (int offset : field_offsets) {
+            object_data_output_.write<byte>(offset);
+        }
+    } else if (data_length < offset_reader::SHORT_OFFSET_READER_RANGE) {
+        for (int offset : field_offsets) {
+            object_data_output_.write<int16_t>(offset);
+        }
+    } else {
+        for (int offset : field_offsets) {
+            object_data_output_.write<int32_t>(offset);
+        }
+    }
+}
+
+void
+default_compact_writer::set_position(const std::string& field_name,
+                                     enum field_kind field_kind)
+{
+    const auto& field_descriptor =
+      check_field_definition(field_name, field_kind);
+    size_t pos = object_data_output_.position();
+    int fieldPosition = pos - data_start_position;
+    int index = field_descriptor.index;
+    field_offsets[index] = fieldPosition;
+}
+
+void
+default_compact_writer::set_position_as_null(const std::string& field_name,
+                                             enum field_kind field_kind)
+{
+    const auto& field_descriptor =
+      check_field_definition(field_name, field_kind);
+    int index = field_descriptor.index;
+    field_offsets[index] = -1;
+}
+
+namespace rabin_finger_print {
+/**
+ * We use uint64_t for computation to match the behaviour of >>> operator
+ * on java. We use >> instead.
+ */
+constexpr uint64_t INIT = 0xc15d213aa4d7a795L;
+
+static std::array<uint64_t, 256>
+init_fp_table()
+{
+    static std::array<uint64_t, 256> FP_TABLE;
+    for (int i = 0; i < 256; ++i) {
+        uint64_t fp = i;
+        for (int j = 0; j < 8; ++j) {
+            fp = (fp >> 1) ^ (INIT & -(fp & 1L));
+        }
+        FP_TABLE[i] = fp;
+    }
+    return FP_TABLE;
+}
+uint64_t
+FP_TABLE_AT(int index)
+{
+    static auto FP_TABLE = init_fp_table();
+    return FP_TABLE[index];
+}
+
+uint64_t
+fingerprint64(uint64_t fp, byte b)
+{
+    return (fp >> 8) ^ FP_TABLE_AT((int)(fp ^ b) & 0xff);
+}
+
+uint64_t
+fingerprint64(uint64_t fp, int v)
+{
+    fp = fingerprint64(fp, (byte)((v)&0xFF));
+    fp = fingerprint64(fp, (byte)((v >> 8) & 0xFF));
+    fp = fingerprint64(fp, (byte)((v >> 16) & 0xFF));
+    fp = fingerprint64(fp, (byte)((v >> 24) & 0xFF));
+    return fp;
+}
+
+uint64_t
+fingerprint64(uint64_t fp, const std::string& value)
+{
+    fp = fingerprint64(fp, (int)value.size());
+    for (const auto& item : value) {
+        fp = fingerprint64(fp, (byte)item);
+    }
+    return fp;
+}
+
+/**
+ * Calculates the fingerprint of the schema from its type name and fields.
+ */
+int64_t
```

Comment:
If we will convert at the end to `int64_t` why we are working on `uint64_t`? I don't know if it has extra cost but at least it looks confusing a little bit

Reply:
` >>` operation works differently on uint64_t and int64_t. To match the behavior of `>>>` (unsigned right shift) on java, I had to work on uint64_t and return int64_t.